### ISSUE TITLE
Does WCAG 4.1.3 require simultaneous delivery of status to sighted and non-sighted users?

### ISSUE THREAD

**Issue by paulbohman** [w3c/wcag#3543]
When a user fills out a form field incorrectly, then leaves the form field (onblur), a common technique is to post an error message immediately on the screen, usually to the right of the field or below the field. This works great for sighted users, but it poses significant usability concerns for screen reader users if the error is announced on blur. The screen reader won't read the error message until after announcing the next element that the user tabbed to. This makes it sound as if the error is associated with the newly focused field, when the error is actually focused with the previous field.

You could say that announcing the error status in real time is bad for screen reader usability.

An alternative method would be to announce the error messages for screen reader users upon arriving at the submit button, before submitting the form. This technique makes the original status messages available to screen reader users in a way that is not confusing, so it is a usability improvement. The user would be able to use shift+tab to go back and fix the errors before submitting the form, and the error messages would be associated correctly with the form fields. The announcement is not simultaneous with the visual display, though, and the question is whether this delayed announcement passes 4.1.3 or not. 

Yet another alternative would be to announce nothing at all until after the form is submitted, then display the error messages. This latter approach would achieve the same goal in the end, but one could argue that it fails 4.1.3 because the original status messages were never announced. 

I created a test page with the first two of these methods (1. aria-live on blur, and 2. error messages associated with the submit button). https://dequeuniversity.com/temp/error-message-on-blur#aria-live-onblur

1. Do both of these examples pass 4.1.3?
2. If only the first example passes, then the "correct" method is also the worst method, in terms of user experience.
3. If both methods pass, then we can agree that even though it would be better to have a way of alerting both sighted and non-sighted users at the same time, it is not required that the alert be made available exactly simultaneously.

This is somewhat related to https://github.com/w3c/wcag/issues/1475 where the question is asked if role=status is required to pass 4.1.3

**Comment by JAWS-test** [comment]
Difficult question that I can't answer.
What would help in any case would be if the name of the incorrect field is specified in the live region, so that it is clear to the screen reader user that it is not the current but the previous field that is incorrect.
The dynamic display of error messages is also not ideal for users of a screen magnifier because they will not notice it. In this respect, it would always be advisable to display and focus error messages after sending, regardless of the error messages before sending.

**Comment by mbgower** [comment]
>1. Do both of these examples pass 4.1.3?
>1. If only the first example passes, then the "correct" method is also the worst method, in terms of user experience.
>1. If both methods pass, then we can agree that even though it would be better to have a way of alerting both sighted and non-sighted users at the same time, it is not required that the alert be made available exactly simultaneously.

1. Yes, in both situations as described, the status messages are surfaced to the screen reader user without receiving focus.
2. They both pass. I don't agree with the statement that the starting design "works great for sighted users", but I agree that there is always a discussion on how to get the best experience for all users which absolutely exceeds the requirements of the baseline set in WCAG. Hopefully the fact there is a requirement to surface status messages is bringing more attention to possible effects on a broader set of users to such interactions.
3. I agree that there is no language in the normative text about the timing of the surfacing of the alert, and in fact it would be impossible to set a timing requirement, given the range of ATs that may be surfacing it, as well as the customization users can set.

**Comment by paulbohman** [comment]
@JAWS-test you make a good point about screen magnifier users, and it sounds like @mbgower you are alluding to the same idea, or at least that the real-time notifications may also pose problems for some sighted users. 

@JAWS-test I sort of like your idea of including the label of the previous field in the real-time aria-live announcement. It would reduce confusion. I say "sort of" because it means that the developer is doing something to make up for a problem in browser and screen reader implementation, so it's a bit of a hack, but it would improve the user experience in most cases. I suppose the developer could also preface the aria-live announcement with something like "Error on the previous field." If you put all of that together, then it would be something like this: 

"Error on the previous field, [field label], [error message]". The preface would have to be worded in a way that took into account that the user could tab forward or backward through the form, so it would be bad to say "Error on the field above," for example.

**Comment by JAWS-test** [comment]
> Error on the previous field, [field label], [error message]

I think: If you have "Error on the previous field" than you don't need "[field label]"

**Comment by JAWS-test** [comment]
> My working assumption (tell me if you agree or not) is that if real-time status messages are shown to sighted users, only announcing an error message on form submit would NOT pass [4.1.3 Status Messages ](https://www.w3.org/TR/WCAG22/#status-messages)because even though the user would receive an error message, the original status message was never announced before form submit.

I think you're right. At least that's how I read the Understanding and the techniques for 4.1.3. 

However, I would consider the violation of 4.1.3 to be minor if error messages are displayed after the submit and are easily recognisable with the screen reader and if error avoidance (before filling in and submitting) is supported by input instructions (SC 3.3.2).

**Comment by mbgower** [comment]
As with a lot of these kinds of discussions, there's always a danger of extrapolating a specific response to a specific use case/scenario and applying it broadly as a general statement. We obviously _do_ need to make sure we have some logical continuum in arguments, but let me lay a bit of groundwork.

### A status message doesn't take focus
> status message
change in content that is not a [change of context](https://www.w3.org/WAI/WCAG21/Understanding/status-messages.html#dfn-changes-of-context), and that provides information to the user on the success or results of an action, on the waiting state of an application, on the progress of a process, or on the existence of errors

This is worth emphasizing because you mention posting "error" messages, and these often take focus -- the user needs to resolve before proceeding -- and thus are not status messages.

### Error messages that do not take focus will often use a have a higher level of assertion than informational messages

I'd argue that none of the messages in the posted examples are error messages. They're just information messages that are conveying the result of how a user interacts with an input. As such, I think there are a lot better ways of designing this interaction which would markedly alter some parts of this discussion.

As well, the verbosity differences are really going to affect when a user receives your messages -- and can lead to some of the considerations already mentioned, even when an overall redesign doesn't take place.

### The scenario got altered?

In the [inital scenario](https://github.com/w3c/wcag/issues/3543#issue-1986983066) #2, you said the status message would be surfaced
> upon arriving at the submit button, before submitting the form

In your [working assumption](https://github.com/w3c/wcag/issues/3543#issuecomment-1806032436), you state
> only announcing an error message on form submit

Those are obviously different scenarios. I just want to be sure we resolve the first discussion before moving on to a 3rd scenario.

Speaking of which, I think that there's nothing to submit in your design because your inputs have already been responded to. So there's really no case for an error message when reaching or activating the submit (and to be clear,  That may be creating unnecessary distraction. If we want to examine this from a closer-to-life perspective, I think it would be useful to make the examples more like a 'normal' set of inputs with some kind of inline error handling. Then we'd be able to assess what happens if a user ignores or misses those prompts and tries to submit. I believe this _could_ meaningfully change the discussion.

**Comment by paulbohman** [comment]
The examples I created may be served better with other models of interaction, that's true, but I was hoping that we could abstract from my examples, and not focus on the actual prompts or responses. The abstraction is that a message is presented to the user upon leaving a form field. That's all. Whether it is an error or an affirmation or an informational notice or success message, or anything else is important on some levels, I agree, but for the sake of the discussion here, what I want to focus on is the fact that a message of some kind visually appears on blur (upon leaving the field).

If the type of message presented happens to be a status message of some kind, then 4.1.3 becomes relevant. Once that is true, then we need to know what methods are acceptable for meeting 4.1.3. Some of the possible techniques include:

Technique A: Use aria-live on blur so the announcement happens "at the same time," but screen reader users will hear the status upon arriving on the next focusable element.

Technique B: Associate the status message with the submit button AND with the individual fields. No real-time on blur announcement.

Technique C: Use aria-live on blur, and add a prefix, such as "The previous field has an error."

Technique D: Don't do anything special to inform screen reader users until the form is submitted, at which time the status message(s) are communicated to the user.

... I'm sure there are other possible techniques, but it would be helpful to get some sort of consensus on whether any or all of these techniques satisfy 4.1.3.

**Comment by patrickhlauke** [comment]
for context, we discussed this in today's WCAG 2.x backlog meeting, and the overall consensus seemed to be that both approaches (at the time there were only two to look at) would pass, as both satisfy the high-level ask of the SC, which in effect is: if somebody can't see the status message, they're not missing out on information. there's no requirement per se that SR users need to have the announcement of the status at the same time as it visually appears, so technique B would work just as well, with the caveat that obviously if the status that is shown is somehow urgent/mission critical then no, delaying its announcement would NOT be equivalent. but in these cases here, that's not a problem

**Comment by paulbohman** [comment]
@patrickhlauke I appreciate the quick response! My own interpretation is in alignment with what you said, though I know others might disagree.

**Comment by KevinP-NZ** [comment]
I agree with @patrickhlauke and @paulbohman  - the key thing is that the messages are conveyed , in full, and have to be resolved at the point at which it matters (before proceeding). Equivalent access is maintained.

**Comment by mbgower** [comment]
It sounds like everyone is in agreement on the initial questions, and the [initial response](https://github.com/w3c/wcag/issues/3543#issuecomment-1805939437).

I suggest we capture that as an addition to the Understanding document. Feel free to submit a draft @paulbohman or anyone!

We _could_ get into quite a few discussions on Paul's [additional scenario](https://github.com/w3c/wcag/issues/3543#issuecomment-1806032436) on post-submit messages, as well as things like the relative merits of "polite" versus "assertive" and how the importance of a status message may affect a user's preferences.

This question and the other issue you referenced got me revisiting Status Messages again. I think it's important to note that a close reading of the normative text suggests a user can meet Status Messages by simply using "aria-describedby", which would be surfaced when focus was put back onto the relevant input (by the system or user). I wouldn't consider that great design on its own, but it seems to meet the normative wording of Status Messages (programmatically determined; the message is surfaced even though it doesn't take focus). It also seems to address one of the concerns raised, which is an overly verbose/disruptive screen reader experience.

I think there is scope for the author to work with users of ATs to come up with preferred, even innovative user experiences, and the Status Messages SC gives some muscle to ensure a more universal experience.

**Comment by mbgower** [comment]
Just following up on this. @paulbohman do you want to do a small PR to update the language of the understanding document to add in some context?
